### PR TITLE
Update provider metadata after adding new connection in browser

### LINKED ISSUE #43382: Adding/remove connections with browser doesn't update provider metadata
**Describe the bug**

When a new connection is added using browser, the provider metadata is not updated. In consequence, the new connection is not listed in the provider metadata. This connection will be available only after relaunching QGIS. 

This is true for GPKG, spatialite, postgres, oracle provider.

In consequence, QgsProviderConnectionComboBox will not list new connections 

Futhermore, when a connection is removed, the provider metadata is not updated too.

**How to Reproduce**

1. List available gpkg connections using the following line in the python command : 

`QgsProviderRegistry.instance().providerMetadata('ogr').connections()`

2. Add a new gpkg connection using the browser

3. List again available gpkg connections. The new connection is missing

When a connection is removed, the connection also stays in the provider metadata

**QGIS and OS versions**

Info | version|
-----|-----------|
QGIS version | 3.19.0-Master | 
QGIS code revision | ddbb0e2e9d
Qt version | 5.12.8
Python version | 3.8.5
GDAL/OGR version | 3.0.4
PROJ version | 6.3.1
EPSG Registry database version | v9.8.6 (2020-01-22)
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
SQLite version | 3.31.1
PostgreSQL client version | 13.3 (Ubuntu 13.3-1.pgdg20.04+1)
SpatiaLite version | 4.3.0a
QWT version | 6.1.4
QScintilla2 version | 2.11.2
OS version | Ubuntu 20.04.2 LTS
This copy of QGIS writes debugging output.
Active Python plugins | |

### PR DESCRIPTION
**Pull request by SebastienPeillet**
## Description

Fix #43382

When a new connection is added using browser, the provider metadata is not updated. In consequence, the new connection is not listed in the provider metadata. This connection will be available only after relaunching QGIS.

This is true for GPKG, spatialite, postgres, oracle provider.

In consequence, QgsProviderConnectionComboBox will not list new connections

Futhermore, when a connection is removed, the provider metadata is not updated too.

Until now the connection was removed only, with the PR QGIS removes the connection from the provider metadata which will remove the connection

### CODE REVIEW

File: src/providers/spatialite/qgsspatialitesourceselect.cpp [hunk 1]
```diff
@@ -312,6 +314,11 @@ bool QgsSpatiaLiteSourceSelect::newConnection( QWidget *parent )
   // inserting this SQLite DB path
   settings.setValue( baseKey + "selected", savedName );
   settings.setValue( baseKey + savedName + "/sqlitepath", myFI.canonicalFilePath() );
+
+  QgsProviderMetadata *providerMetadata = QgsProviderRegistry::instance()->providerMetadata( QStringLiteral( "spatialite" ) );
+  QgsSpatiaLiteProviderConnection *providerConnection =  static_cast<QgsSpatiaLiteProviderConnection *>( providerMetadata->createConnection( savedName ) );
+  providerMetadata->saveConnection( providerConnection, savedName );
+
```
Comment (nyalldawson) [inline src/providers/spatialite/qgsspatialitesourceselect.cpp:321]: won't the call to QgsSpatiaLiteConnection::deleteConnection in btnDelete_clicked also need to be fixed? (and then QgsSpatiaLiteConnection::deleteConnection can be totally removed -- it's not part of stable api, and will be unused code)

File: src/providers/spatialite/qgsspatialitesourceselect.cpp [hunk 2]
```diff
@@ -312,6 +314,11 @@ bool QgsSpatiaLiteSourceSelect::newConnection( QWidget *parent )
   // inserting this SQLite DB path
   settings.setValue( baseKey + "selected", savedName );
   settings.setValue( baseKey + savedName + "/sqlitepath", myFI.canonicalFilePath() );
```
Comment (nyalldawson) [inline src/providers/spatialite/qgsspatialitesourceselect.cpp:316]: The settings part is no longer required, right? It's already handled in QgsSpatiaLiteProviderConnection::store which will be called by the call to saveConnection below.

File: src/providers/spatialite/qgsspatialitedataitemguiprovider.cpp [hunk 3]
```diff
@@ -134,7 +136,9 @@ void QgsSpatiaLiteDataItemGuiProvider::deleteConnection( QgsDataItem *item )
                               QMessageBox::Yes | QMessageBox::No, QMessageBox::No ) != QMessageBox::Yes )
     return;
 
-  QgsSpatiaLiteConnection::deleteConnection( item->name() );
+  QgsProviderMetadata *providerMetadata = QgsProviderRegistry::instance()->providerMetadata( QStringLiteral( "spatialite" ) );
+  providerMetadata->deleteConnection( item->name() );
+
```
Comment (nyalldawson) [inline src/providers/spatialite/qgsspatialitedataitemguiprovider.cpp:141]: QgsSpatiaLiteDataItemGuiProvider::createDatabase should be updated to use providerMetadata->saveConnection instead too

File: src/core/providers/ogr/qgsogrdataitems.cpp [hunk 4]
```diff
@@ -436,6 +437,9 @@ bool QgsOgrDataCollectionItem::saveConnection( const QString &path, const QStrin
       QgsOgrDbConnection connection( connName, ogrDriverName );
       connection.setPath( path );
       connection.save();
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrdataitems.cpp:439]: is the QgsOgrDbConnection code here redundant now?

File: src/core/providers/ogr/qgsogrdbconnection.cpp [hunk 5]
```diff
@@ -89,6 +91,9 @@ void QgsOgrDbConnection::deleteConnection( const QString &connName, const QStrin
 {
   QgsSettings settings;
   settings.remove( QStringLiteral( "%1/%2" ).arg( connectionsPath( settingsKey ), connName ) );
```
Comment (nyalldawson) [inline src/core/providers/ogr/qgsogrdbconnection.cpp:93]: This looks redundant

File: src/providers/oracle/qgsoracleconn.cpp [hunk 6]
```diff
@@ -24,6 +24,8 @@
 #include "qgsoracletablemodel.h"
 #include "qgssettings.h"
 #include "qgsoracleconnpool.h"
+#include "qgsprovidermetadata.h"
+#include "qgsproviderregistry.h"
```
Comment (nyalldawson) [inline src/providers/oracle/qgsoracleconn.cpp:28]: not required?

File: src/providers/oracle/qgsoracledataitems.cpp [hunk 7]
```diff
@@ -313,7 +313,8 @@ void QgsOracleConnectionItem::deleteConnection()
                               QMessageBox::Yes | QMessageBox::No, QMessageBox::No ) != QMessageBox::Yes )
     return;
 
-  QgsOracleConn::deleteConnection( mName );
+  QgsProviderMetadata *providerMetadata = QgsProviderRegistry::instance()->providerMetadata( QStringLiteral( "oracle" ) );
+  providerMetadata->deleteConnection( mName );
```
Comment (nyalldawson) [inline src/providers/oracle/qgsoracledataitems.cpp:317]: Looks like QgsOracleSourceSelect::on_btnDelete_clicked() also needs updating

**Comment by SebastienPeillet** [comment]
@nyalldawson indeed I mostly pay attention on the fact that connections were saved in the metadata provider, but it is redundant with QgsSettings :+1: I removed the affected lines.

File: src/providers/oracle/qgsoracleconn.cpp (hunk 6 above)
Comment (SebastienPeillet) [reply src/providers/oracle/qgsoracleconn.cpp]: no, I remove it

File: src/providers/spatialite/qgsspatialitesourceselect.cpp (hunk 1 above)
Comment (SebastienPeillet) [reply src/providers/spatialite/qgsspatialitesourceselect.cpp:318]: Indeed, I removed it after the changes on btnDelete_clicked

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**